### PR TITLE
	Fix reuse update method bug

### PR DESCRIPTION
**Pull request by jakwuh**

### CODE REVIEW

**Comment by jakwuh** [comment]
This test case should be passed but it fails. Let me give some explanations.
When we use following definition:

``` js
A: {
    test: 'Test'
},

A1: '!A#a',
A2: '!A#b'
```

then we expect A1 and A2 to reuse one instance A, but to call different update method on it. In fact, as far as `A#a` and `A#b` are different moduleId's in `di.js`, we result into different instances.

I think this should be fixed.

**Comment by jakwuh** [comment]
@ftdebugger if you see the problem and agree with it, let me know so I will try to fix it.

**Comment by ftdebugger** [comment]
👍

**Comment by jakwuh** [comment]
@ftdebugger fixed. Please take a short look cause I need this pr to be merged & published

File: index.js [hunk 1]
```diff
@@ -439,9 +439,20 @@ let parseStringDefinition = (definition) => {
     }
 
     if (definition[0] === '!') {
-        return {
-            reuse: definition.slice(1)
+        let {bundleName, factory, update} = parseStringDefinition(definition.slice(1));
```
Comment (ftdebugger) [inline index.js]: Definition already parsed:

``` js
let bundleName: matches[1],
    factory: matches[3],
    update: matches[5];
```

File: index.js [hunk 2]
```diff
@@ -439,9 +439,20 @@ let parseStringDefinition = (definition) => {
...
+        let reuse;
+        if (factory) {
+            reuse = `${bundleName}.${factory}`;
```
Comment (ftdebugger) [inline index.js]: Looks like inconsistent here, you can override update method, but not factory. How about allow define factory for each reuse too? It is make sense, when you need different construct strategy for each reuse at first use.

I think `bundleName` is enough.

``` js
    let matches = definition ?
        definition.match(/^!?([^.#]+)(\.([^#]+))?(#(.+))?$/) :
        null;

    if (!matches) {
        throw new Error('Unknown module format: ' + JSON.stringify(definition));
    }

    let bundleName = matches[1],
        factory = matches[3],
        update = matches[5];

    if (definition[0] === '!') {
        let reuseDefinition = {
            reuse: bundleName
        };

        if (factory) {
            reuseDefinition.factory = factory;
        }

        if (update) {
            reuseDefinition.update = update;
        }

        return reuseDefinition;
    } else {
        return {
            parentId: definition,
            bundleName,
            factory,
            update
        };
    }
```

**Review by ftdebugger — CHANGES_REQUESTED** [review]
Check my notes

**Review by ftdebugger — APPROVED** [review]

**Pull request merged**